bindings/src/main/java/io/pravega/storage/filesystem/FileSystemChunkStorage.java

```diff
@@ -0,0 +1,294 @@
+            int totalBytesRead = 0;
+            long readOffset = fromOffset;
+            do {
+                ByteBuffer readBuffer = ByteBuffer.wrap(buffer, bufferOffset, length);
+                int bytesRead = channel.read(readBuffer, readOffset);
+                bufferOffset += bytesRead;
+                totalBytesRead += bytesRead;
+                length -= bytesRead;
+                readOffset += bytesRead;
+            } while (length > 0);
+            return totalBytesRead;
+        } catch (IOException e) {
+            throw convertExeption(handle.getChunkName(), "doRead", e);
+        }
+    }
+
+    @Override
+    protected int doWrite(ChunkHandle handle, long offset, int length, InputStream data) throws ChunkStorageException {
+        Path path = Paths.get(config.getRoot(), handle.getChunkName());
+
+        long totalBytesWritten = 0;
+        try (FileChannel channel = getFileChannel(path, StandardOpenOption.WRITE)) {
+            long fileSize = channel.size();
+            if (fileSize != offset) {
+                throw new IndexOutOfBoundsException(String.format("fileSize (%d) did not match offset (%d) for chunk %s", fileSize, offset, handle.getChunkName()));
+            }
+
+            // Wrap the input data into a ReadableByteChannel, but do not close it. Doing so will result in closing
+            // the underlying InputStream, which is not desirable if it is to be reused.
+            ReadableByteChannel sourceChannel = Channels.newChannel(data);
+            while (length > 0) {
+                long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                assert bytesWritten > 0 : "Unable to make any progress transferring data.";
+                offset += bytesWritten;
+                totalBytesWritten += bytesWritten;
+                length -= bytesWritten;
+            }
+            channel.force(true);
+        } catch (IOException e) {
+            throw convertExeption(handle.getChunkName(), "doWrite", e);
+        }
+        return (int) totalBytesWritten;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException {
+        try {
+            int totalBytesConcated = 0;
+            Path targetPath = Paths.get(config.getRoot(), chunks[0].getName());
+            long offset = chunks[0].getLength();
+
+            for (int i = 1; i < chunks.length; i++) {
+                val source = chunks[i];
+                Preconditions.checkArgument(!chunks[0].getName().equals(source.getName()), "target and source can not be same.");
+                Path sourcePath = Paths.get(config.getRoot(), source.getName());
+                long length = chunks[i].getLength();
+                Preconditions.checkState(offset <= getFileSize(targetPath));
+                Preconditions.checkState(length <= getFileSize(sourcePath));
+                try (FileChannel targetChannel = getFileChannel(targetPath, StandardOpenOption.WRITE);
```

Comment:
why are you opening the same file more times ?
it is better to open the file for write only once.

I suggest you also to:
- create a temporary file
- fill it
- rename new the file with ATOMIC_MOVE flag to the targetPath
This way we are protected from partial writes

Reply:
Good catch that I was opening same file multiple times. Fixing it.

the target chunk could be multi GB file.

As for why partial writes are still okay...
- chunk metadata is not updated unless the whole operation succeeds. So even if resultant chunk has more data we never consider it valid part of chunk.
- The source chunks are not deleted.

Reply:
Fixed for loop